File: clients/gui/src/host/elements/score.rs
//! `score` — an engraved notation page: the leaf whose state is a **document
//! it does not own**.
//!
//! The host holds geometry, never the score: a client engraves and sends a
//! semantic display list, and everything here is a *reading* of that drawing —
//! which element is under the cursor, how far up the staff a drag has moved it,
//! where the playback cursor sits at a given millisecond. The page itself and
//! all of that reading stay in [`crate::host::score`], which is the model; this
//! file is only how the passes reach it.
//!
//! What the port collapses is the routing that model needed while the leaf was
//! an enum arm: four `interact` doors (`score_select`, `score_drag`,
//! `score_drag_end`, plus the payload readers), a `nav::score_steps` lookup and
//! a `Drag::ScoreStep` variant carrying the press-time origin the machine could
//! not put anywhere else. The press-time origin is a field here, and the drag's
//! displacement was always the page's own (`ScoreData::drag`), drawn as
//! notation while it happens.
//!
//! **An edit is an intent, and the preview outlives the gesture.** The release
//! reports `"transpose" <xml:id> <steps>` — the owner's units, not pixels — and
//! the displacement **stays drawn** until the client answers with a re-engraved
//! page, because dropping it first would show the old pitch for a frame. That
//! is why `display_list` replaces the drawing and keeps the chrome.

use serde_json::{Map, Value};

use clausters_core::osc::OscType;

use crate::host::paint::Draw;
use crate::host::score::{ScoreColors, ScoreData, ScoreDrag};
use crate::host::widget::element::{Claim, Ctx, Element, Events, Input, Needs};
use crate::host::widget::parse;

/// An engraved page, plus the one thing a page does not carry: where the pitch
/// drag in flight was pressed.
#[derive(Debug, Clone)]
pub struct Score {
    pub data: ScoreData,
    /// The press this drag started from, in window pixels — the origin the
    /// step count is measured from, so it is absolute from the snapshot rather
    /// than accumulated. `None` when no drag is in flight.
    origin_y: Option<f64>,
}

pub(super) fn build(
    props: &Map<String, Value>,
    _blobs: &[Vec<u8>],
) -> Result<Box<dyn Element>, String> {
    Ok(Box::new(Score {
        data: ScoreData::parse(props),
        origin_y: None,
    }))
}

impl Element for Score {
    fn set(&mut self, key: &str, v: &Value) -> bool {
        let data = &mut self.data;
        match key {
            // Replace the engraved page in place — the answer to an edit, and
            // the reason a score does not have to be redefined to change. Only
            // the drawing travels: the chrome (playhead, selection) is the
            // host's own state and survives, so the note the user is editing
            // stays selected across the round trip. The drag preview is what
            // this page *is* now, so it retires here.
            "display_list" => match parse::as_props(v) {
                Some(props) => {
                    let page = ScoreData::parse(&props);
                    let keep = std::mem::replace(data, page);
                    data.playhead = keep.playhead;
                    data.playhead_at = keep.playhead_at;
                    data.playhead_loop_start = keep.playhead_loop_start;
                    data.playhead_loop_len = keep.playhead_loop_len;
                    data.sample_rate = keep.sample_rate;
                    data.selected = keep.selected;
                    // A re-engraved page carries only the drawing; whether the
                    // widget edits is the host's own state, like the chrome, so
                    // an editor stays an editor across the round trip.
                    data.editable = keep.editable;
                    true
                }
                None => false,
            },
            // Locate the static playback cursor; a negative time hides it.
            "playhead" => v.as_f64().map(|t| data.playhead = t as f32).is_some(),
            // Anchor score time 0 to a sample-clock value: the cursor then
            // sweeps on its own, one message per pass instead of per frame.
            "playhead_at" => v.as_f64().map(|t| data.playhead_at = t).is_some(),
            // Wrap the sweep inside a repeated passage (ms; <= 0 length = the
            // straight pass).
            "playhead_loop_start" => v
                .as_f64()
                .map(|t| data.playhead_loop_start = t as f32)
                .is_some(),
            "playhead_loop_len" => v
                .as_f64()
                .map(|t| data.playhead_loop_len = t as f32)
                .is_some(),
            "sample_rate" => v.as_f64().map(|r| data.sample_rate = r).is_some(),
            // Select an element by its MEI id; the empty string clears it.
            "selected" => v
                .as_str()
                .map(|s| data.selected = (!s.is_empty()).then(|| s.to_string()))
                .is_some(),
            // Turn editing on or off live (a view that becomes an editor, or
            // the reverse). A drag only transposes while this is true.
            "editable" => v.as_bool().map(|b| data.editable = b).is_some(),
            _ => false,
        }
    }

    fn draw(&self, d: &mut Draw, ctx: &Ctx) {
        // Notation tessellates straight into the shared triangle mesh: a paper
        // panel under the engraving, glyphs and fills in ink, the playback
        // cursor over it in the playhead accent.
        let (mesh, _, theme) = d.parts();
        mesh.rect(ctx.rect, theme.panel);
        // The cursor sweeps off the engine clock while a pass plays
        // (`playhead_at`), so playback costs no messages per frame.
        let head = self
            .data
            .head_ms(ctx.world.sample_clock, ctx.world.sample_rate);
        let colors = ScoreColors {
            ink: theme.text,
            playhead: theme.playhead,
            selection: theme.selection,
        };
        self.data.render(mesh, ctx.rect, ctx.clip, head, colors);
    }

    fn needs(&self) -> Needs {
        Needs {
            // A score carries its **own** playhead anchor rather than a
            // navigation group's, so it is the widget itself that says its
            // cursor is sweeping — without this the window stops following the
            // clock and the cursor freezes where it was anchored.
            clock: self.data.playhead_at >= 0.0,
            ..Needs::default()
        }
    }

    fn info(&self) -> Vec<(String, Value)> {
        // The one prop a gesture changes: a click selects, and `/gui_set
        // selected` is how a script would reproduce it.
        vec![(
            "selected".into(),
            Value::from(self.data.selected.clone().unwrap_or_default()),
        )]
    }

    fn press(&mut self, at: (f64, f64), input: &Input) -> Claim {
        // A press names the engraved element under it by its MEI id — the same
        // id the client engraved from, so a driver resolves it in its own
        // score. Pressing blank paper clears the selection.
        let picked = self
            .data
            .hit(input.rect, at.0 as f32, at.1 as f32)
            .map(str::to_string);
        let changed = picked != self.data.selected;
        if changed {
            self.data.selected = picked.clone();
        }
        // ...and, on an editable score, holding it drags the element's pitch. A
        // press that does not move stays a plain selection: the release emits
        // nothing more. A read-only page (the default) still selects and
        // reports the element above, but a drag does nothing — the host holds
        // no score, so an edit the client will not apply is a gesture it cannot
        // fulfil.
        let dragging = self.data.editable && picked.is_some();
        if dragging {
            self.data.drag = Some(ScoreDrag {
                id: picked.clone().unwrap_or_default(),
                steps: 0,
            });
            self.origin_y = Some(at.1);
        }
        match (changed, dragging) {
            // Nothing selected, nothing to drag: the press was never this
            // page's, so it goes back to the chain.
            (false, false) => Claim::Decline,
            (true, _) => Claim::events(Events::message(vec![
                OscType::String("element".into()),
                OscType::String(picked.unwrap_or_default()),
            ])),
            (false, true) => Claim::take(),
        }
    }

    fn drag(&mut self, at: (f64, f64), input: &Input) -> Events {
        // Absolute from the press, quantized to whole steps: the page is
        // redrawn only when the drag crosses one, so the pixels between two
        // pitches cost nothing. Nothing is reported until the release — what
        // travels is the finished intent.
        let Some(origin_y) = self.origin_y else {
            return Events::none();
        };
        let steps = self.data.steps_for(input.rect, (at.1 - origin_y) as f32);
        if let Some(drag) = self.data.drag.as_mut() {
            drag.steps = steps;
        }
        Events::none()
    }

    fn release(&mut self, _at: (f64, f64), _input: &Input) -> Events {
        self.origin_y = None;
        let Some(drag) = self.data.drag.as_ref() else {
            return Events::none();
        };
        // A drag that ended where it started retires here — there is nothing to
        // ask the client for. One that moved **keeps its displacement drawn**:
        // the host owns no notation, so it cannot re-engrave the page itself,
        // and dropping the preview now would show the old pitch until the
        // client's answer arrives. The page it sends back retires the preview
        // (see the `display_list` prop).
        if drag.steps == 0 {
            self.data.drag = None;
            return Events::none();
        }
        Events::message(vec![
            OscType::String("transpose".into()),
            OscType::String(drag.id.clone()),
            OscType::Int(drag.steps),
        ])
    }

    fn clone_box(&self) -> Box<dyn Element> {
        Box::new(self.clone())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::host::layout::Rect;
    use crate::host::metrics::Metrics;
    use crate::host::widget::element::Mods;

    /// A one-staff page with two identified noteheads, engraved the way the
    /// client sends one: a viewBox, one glyph outline, and placed primitives.
    fn page(editable: bool) -> Score {
        let props: Map<String, Value> = serde_json::from_str(&format!(
            r#"{{"vb":[1000,400],"step":90,"editable":{editable},
                "glyphs":{{"E0A4":"M0 0 L100 0 L100 -100 L0 -100 Z"}},
                "prims":[
                  {{"k":"glyph","cp":"E0A4","xf":[100,200,1,-1],"id":"n1"}},
                  {{"k":"glyph","cp":"E0A4","xf":[400,200,1,-1],"id":"n2"}}]}}"#
        ))
        .unwrap();
        Score {
            data: ScoreData::parse(&props),
            origin_y: None,
        }
    }

    fn input<'a>(m: &'a Metrics) -> Input<'a> {
        Input {
            metrics: m,
            indent: 0.0,
            rect: Rect::new(0.0, 0.0, 500.0, 200.0),
            scale: 1.0,
            mods: Mods::default(),
            viewport: (600.0, 400.0),
            time: None,
        }
    }

    /// The window pixel a page point lands on, through the same fit the
    /// renderer draws with — so a test presses where the ink is.
    fn at(score: &Score, rect: Rect, px: f32, py: f32) -> (f64, f64) {
        let fit = score.data.fit(rect);
        let [x, y] = fit.apply(px, py);
        (x as f64, y as f64)
    }

    /// A click names the element under it and clears on blank paper — the
    /// inspection half, which is **not** gated by `editable`.
    #[test]
    fn a_press_selects_the_element_under_it() {
        let metrics = Metrics::default();
        let input = input(&metrics);
        let mut score = page(false);

        let claim = score.press(at(&score, input.rect, 150.0, 250.0), &input);
        assert_eq!(
            claim,
            Claim::events(Events::message(vec![
                OscType::String("element".into()),
                OscType::String("n1".into()),
            ]))
        );
        assert_eq!(score.data.selected.as_deref(), Some("n1"));
        assert_eq!(
            score.info(),
            vec![("selected".into(), Value::from("n1"))],
            "a query answers what is selected now"
        );

        // Blank paper clears it, and says so.
        let claim = score.press(at(&score, input.rect, 900.0, 380.0), &input);
        assert_eq!(
            claim,
            Claim::events(Events::message(vec![
                OscType::String("element".into()),
                OscType::String(String::new()),
            ]))
        );
        assert_eq!(score.data.selected, None);
        // ...and pressing blank paper *again* changes nothing, so the press
        // goes back to the chain instead of being swallowed.
        assert_eq!(
            score.press(at(&score, input.rect, 900.0, 380.0), &input),
            Claim::Decline
        );
    }

    /// A read-only page selects and reports, and drags nothing: the host holds
    /// no score, so an edit the client will not apply is a gesture it cannot
    /// fulfil.
    #[test]
    fn a_read_only_page_does_not_drag() {
        let metrics = Metrics::default();
        let input = input(&metrics);
        let mut score = page(false);
        score.press(at(&score, input.rect, 150.0, 250.0), &input);
        assert!(score.data.drag.is_none());
        let (x, y) = at(&score, input.rect, 150.0, 250.0);
        score.drag((x, y - 100.0), &input);
        assert!(score.data.drag.is_none());
        assert_eq!(score.release((x, y - 100.0), &input), Events::none());
    }

    /// An editable page displaces the element as the drag crosses whole
    /// diatonic steps, and the release reports the intent in the owner's units
    /// — steps up the staff, never pixels.
    #[test]
    fn an_editable_page_transposes_in_whole_steps() {
        let metrics = Metrics::default();
        let input = input(&metrics);
        let mut score = page(true);
        let press = at(&score, input.rect, 150.0, 250.0);
        score.press(press, &input);
        assert_eq!(score.data.drag.as_ref().map(|d| d.steps), Some(0));

        // Two steps up, in page units through the fit: dragging up is positive.
        let step_px = (score.data.step * score.data.fit(input.rect).sy) as f64;
        score.drag((press.0, press.1 - 2.0 * step_px), &input);
        assert_eq!(score.data.drag.as_ref().map(|d| d.steps), Some(2));

        let events = score.release((press.0, press.1 - 2.0 * step_px), &input);
        assert_eq!(
            events,
            Events::message(vec![
                OscType::String("transpose".into()),
                OscType::String("n1".into()),
                OscType::Int(2),
            ])
        );
        assert!(
            score.data.drag.is_some(),
            "the displacement stays drawn until the client's re-engraved page arrives"
        );
    }

    /// A drag that ends where it started asks for nothing and drops its
    /// preview — the press was a selection after all.
    #[test]
    fn a_drag_that_moved_nothing_retires_on_release() {
        let metrics = Metrics::default();
        let input = input(&metrics);
        let mut score = page(true);
        let press = at(&score, input.rect, 150.0, 250.0);
        score.press(press, &input);
        assert_eq!(score.release(press, &input), Events::none());
        assert!(score.data.drag.is_none());
    }

    /// A swept page declares that it reads the clock and a still one does not
    /// — the declaration that used to be a `live.rs` arm asking what kind of
    /// widget this was.
    #[test]
    fn a_swept_page_declares_that_it_reads_the_clock() {
        let mut score = page(false);
        assert!(!score.needs().clock);
        assert!(score.set("playhead_at", &Value::from(48_000.0)));
        assert!(score.needs().clock);
    }

    /// A re-engraved page keeps the host's own chrome — the selection the user
    /// is editing, the playhead — and retires the drag preview it answers.
    #[test]
    fn a_new_display_list_keeps_the_chrome_and_retires_the_preview() {
        let mut score = page(true);
        score.data.selected = Some("n2".into());
        score.data.playhead = 500.0;
        score.data.drag = Some(ScoreDrag {
            id: "n2".into(),
            steps: 3,
        });
        assert!(score.set(
            "display_list",
            &Value::from(r#"{"vb":[1000,400],"prims":[]}"#)
        ));
        assert_eq!(score.data.selected.as_deref(), Some("n2"));
        assert_eq!(score.data.playhead, 500.0);
        assert!(score.data.editable, "an editor stays an editor");
        assert!(score.data.drag.is_none());
        assert!(score.data.prims.is_empty(), "and the drawing was replaced");
    }
}

File: clients/python/examples/notation/score.py
#!/usr/bin/env python3
"""Engraving music notation into the GUI host: the ``score`` widget.

A read-only view like ``plot`` and the node tree, but of a **musical score**
rather than a signal. The client engraves a score with verovio (bundled in the
package) into a semantic display list -- a SMuFL glyph-outline table plus
placed glyphs, staff lines, stems and beams in page units -- and the host
tessellates it into the same triangle mesh the rest of the chrome uses. verovio
lives entirely on the client side; the host never depends on it.

One engraving carries three layers: what is **drawn**, where the **playback
cursor** sits at each onset, and the **notes** that sound. This example places
the notes on a timeline and drives them from a **transport bar** -- play, pause,
stop, rewind and play-from-the-selected-note -- anchoring the cursor to the
engine's sample clock, so the score follows the audio with **one message per
pass** (``playhead_at``), the host reading the clock every frame from there,
exactly as the timeline views do. A stopped transport is the other half of that
one number: it goes negative and the static ``playhead`` holds the cursor where
the music was left. It is the **shared** transport (`clausters.gui.Transport`),
the same object the multitrack editor drives its lanes with: a page differs only
in the unit its static cursor is placed in, and that is all
`notation.transport` fills in.

The page is also **clickable and editable**: every primitive carries the MEI
``xml:id`` it was engraved from, so a press reports the element under the cursor
as an ``"element"`` event and the host highlights it, and dragging one up or
down the staff reports a ``"transpose"`` naming the diatonic staff position that
element **reaches** -- absolute rather than a displacement, so a resend cannot
move it twice. Because the id is the client's own, this script resolves both against its
own score: a click sounds the note, a drag transposes it, re-engraves the page
and sends it back -- the whole edit round trip, with nothing shared but the id.

The engraver is **libverovio**, which ships inside the installed package -- like
the Faust compiler, nothing to install separately. In a source checkout, build
and stage it once (``third_party/BUILD-VEROVIO.md``)::

    third_party/build-verovio.sh
    python clients/python/build_native.py

Then, with the client importable (``pip install ./clients/python`` or
``PYTHONPATH=clients/python``)::

    python clients/python/examples/notation/score.py

A window opens showing the engraved phrase, stopped at the top -- press **play**
and the cursor follows the sound. Click a note to hear it and select it, drag one
up or down to transpose it, and **from note** plays from the one selected;
**undo**/**redo** walk the edits. Close the window to stop. Needs an audio
device, a display and a GPU adapter.

The undo stack is the client's, not the host's: the host holds no score, so
every edit -- and every step back through it -- happens on this side, against
`clausters.gui.notation.Score`. The host only draws the page it is sent.

This file is organized as ``# %%`` cells (the VS Code / Jupyter convention):
step through it with Shift+Enter and the window stays up between cells, or run
it as a plain script.
"""

# %%
import sys

from clausters import Event, Session, play
from clausters.gui import button, notation, panel, source, view
from clausters.seq.timeline import Playhead, Timeline

# Six bars in ABC -- the readable way to type a score by hand; verovio reads MEI
# and MusicXML through the same loader, which is what a score usually arrives as.
# The header is the whole grammar you need here: `M:` the meter, `L:` the length
# a bare letter means (a quarter), `K:` the key (G, so every F is sharp). Then a
# letter is a note (`C` is middle C, `c` the one above), `/` halves it and a
# digit multiplies it, `[CEG]` is a chord, and `|` bars it. Each bar fills its
# 4/4 exactly -- verovio drops what overflows a measure, so an over-full bar
# would be drawn short and sound short.
PHRASE = """X:1
T:Six bars
M:4/4
L:1/4
K:G
C D E F | G/A/G/F/ E D | C D/E/F/G/ A | G2 F E | [CEG] G C2 | C4 |
"""

# One beat per second, so the engraving's milliseconds are beats/1000: score time
# and clock time become the same axis, which is what lets one anchor tie the
# cursor to the sound.
TEMPO = 1.0

# %% [markdown]
# ## The window

# %%
def scene(engraved, sample_rate: float) -> dict:
    """A transport bar over a scrollable, zoomable view of the engraved score.

    Every widget is *named* — the seven transport buttons and the ``score`` page
    — so the script drives each by name and never picks an id. The bar is chrome:
    a fixed height, so the page takes all the rest however the window is
    resized.

    ``engraved`` is a `clausters.gui.source` holding the display list, not the
    list itself: an edit re-engraves the score and calls ``engraved.set(...)``,
    which
    reaches this definition *and* every window drawing it. Handed the raw dict
    the view would draw the same page and the definition would go stale the
    first time a note moved."""
    return view(
        panel(button(name="rewind", label="|<"),
              button(name="play", label="play"),
              button(name="pause", label="pause"),
              button(name="stop", label="stop"),
              button(name="from_note", label="from note"),
              button(name="undo", label="undo"),
              button(name="redo", label="redo"),
              layout="row", h=34.0),
        notation.score_view(engraved, name="score",
                            width=880.0, sample_rate=sample_rate, editable=True),
        layout="col", title="Engraved score (verovio -> GPU)", w=920, h=420,
    )


# %% [markdown]
# ## What the page plays back

# %%
def phrase_timeline(notes: list) -> Timeline:
    """Place the engraved notes on a `Timeline` (see `TEMPO`: score ms are
    beats/1000). Built per play, so a transposed note is played at the pitch it
    now has."""
    timeline = Timeline()
    for note in notes:
        timeline.add(note["t"] / 1000.0,
                     Event(midinote=note["pitch"], dur=note["dur"] / 1000.0,
                           amp=0.12))
    return timeline


# %% [markdown]
# ## Engrave, open, and wire it up

# %%
# `Score` rather than `engrave`: it keeps the document open, so the page the
# window shows can be edited and re-engraved against it (a narrow page, so
# the phrase wraps into a few systems and the view scrolls).
score = notation.Score(PHRASE, page_width=1100)
dl = score.display_list()
print(f"engraved: {len(dl['glyphs'])} glyph outlines, "
      f"{len(dl['prims'])} primitives, {len(dl['cursors'])} cursor stops, "
      f"{len(dl['notes'])} notes, page {dl['vb']}")

# the session is the ambient one for the whole block, so a bare `play` below
# resolves to its server and clock
session = Session.live(tempo=TEMPO)
server = session.server
# `query_info` rather than the launch options: it is the one spelling both
# clients have, so this file and its page twin ask the same question.
sr = server.query_info().nominal_sample_rate
gui = session.gui()
engraved = source(display_list=dl)
win = scene(engraved, sr).open()

session.start()                     # the clock runs the routines

# Both round trips run off the same id: the widget reports the MEI id
# under the cursor, and that id indexes this script's own engraving.
by_id = {note["id"]: note for note in dl["notes"]}
selected = None


def pass_from(at):
    """One playback pass: the engraved notes on a fresh `Timeline`, played
    from beat `at`. The transport calls this on every play, so a note
    transposed meanwhile simply sounds at the pitch it now has."""
    return Playhead(phrase_timeline(dl["notes"]), session.clock,
                    server).play(at=at)

def phrase_end():
    """Where the piece ends, in beats: the last note's onset plus its
    length. The transport parks the cursor there when a pass runs out."""
    last = dl["notes"][-1]
    return (last["t"] + last["dur"]) / 1000.0

# The transport is the shared one (`clausters.gui.Transport`), the same
# object the multitrack editor drives its lanes with; `notation.transport`
# only fills in the page's unit -- a score cursor is placed in score
# milliseconds, not samples.
transport = notation.transport(gui, win["score"].id, source=pass_from,
                               tempo=TEMPO, sample_rate=sr, extent=phrase_end)
transport.locate(0.0)               # the cursor waits at the top
print("press play -- click a note to hear it and to select it, drag one "
      "up or down to transpose it, 'from note' plays from the selected "
      "one, undo/redo walk the edits; close the window to stop")

def from_note():
    """Play from the selected note: the click round trip put its id in
    `selected`, and this script's own engraving says when it sounds."""
    if selected not in by_id:
        print("  no note selected: click one first")
        return
    transport.play(server, at=by_id[selected]["t"] / 1000.0)

def refresh_page():
    """Re-engrave the score and replace the drawn page in place, rebuilding
    the id index. Every edit ends here -- a drag, an undo, a redo -- since
    all the host needs is the new display list; the host keeps the playhead
    and selection across the swap.

    ``engraved.set`` is the whole of it: the source rewrites the definition and
    pushes the drawing layers to every window already showing them, so a second
    window (or a re-open of this one) shows the score as edited rather than as
    engraved."""
    global dl, by_id
    dl = score.display_list()
    engraved.set(dl)
    by_id = {note["id"]: note for note in dl["notes"]}

def undo():
    """Step back one edit. `Score` owns the undo stack (a stack of MEI
    snapshots) -- the host holds no score, so undo is the client's, and it
    answers False rather than crashing when there is nothing to undo."""
    if score.undo():
        refresh_page()
        print("  undo")
    else:
        print("  nothing to undo")

def redo():
    if score.redo():
        refresh_page()
        print("  redo")
    else:
        print("  nothing to redo")

def on_score(tag, *payload):
    """The page's two edit-backs, wired to the ``score`` handle. A click
    reports the MEI id under the cursor (``"element"``) — this side
    selects it and sounds it; a drag reports a ``"transpose"`` naming the
    staff position reached — this side makes it true, re-engraves and sends the
    page back. The handlers read `dl`/`by_id` when they run, so an edit
    made meanwhile is simply played. An event handler runs on the client's
    reply thread, and the ambient session is per-thread, so every `play`
    here names its `server` instead of letting it resolve."""
    global selected
    if tag == "element" and payload:
        selected = payload[0] or None
        note = by_id.get(payload[0])
        if note is None:
            print(f"  clicked {payload[0] or '(blank paper)'}")
            return
        print(f"  clicked note {payload[0]}: MIDI {note['pitch']} "
              f"at {note['t']:.0f} ms")
        play(Event(midinote=note["pitch"], dur=note["dur"] / 1000.0,
                   amp=0.15), server=server)
    elif tag == "transpose" and len(payload) >= 2:
        # The host drew the drag; this side makes it true -- move the note
        # **to** the staff position the payload names, re-engrave, and send
        # the page back, which is what retires the preview. The ids survive
        # the edit, so `by_id` keeps indexing the same notes (at their new
        # pitches) and the note stays selected.
        #
        # The payload is a position and not a displacement, which is what
        # makes it safe to arrive late or twice: `transpose_to` computes the
        # step count against the engraving it has right now, so a page this
        # side re-engraved meanwhile cannot make the edit land somewhere else.
        element, position = payload[0], int(payload[1])
        if not score.transpose_to(element, position):
            print(f"  refused to transpose {element}: this verovio has "
                  "no working editor")
            return
        refresh_page()
        note = by_id.get(element)
        print(f"  moved {element} to staff position {position:+d}"
              + (f" -> MIDI {note['pitch']}" if note else ""))
        if note is not None:
            play(Event(midinote=note["pitch"], dur=note["dur"] / 1000.0,
                       amp=0.15), server=server)

# Wire every widget by name: each button acts on its **click** -- the press
# completed on the button, so sliding off before letting go cancels it -- and
# `locate` doubles as rewind, so playing it starts a fresh pass from the top.
# The score page answers its own edit-backs.
win["play"].on_click(lambda: transport.play(server))
win["pause"].on_click(transport.pause)
win["stop"].on_click(transport.stop)
win["rewind"].on_click(lambda: transport.locate(0.0))
win["from_note"].on_click(from_note)
win["undo"].on_click(undo)
win["redo"].on_click(redo)
win["score"].on_event(on_score)
win.on_closed(lambda: print("window closed"))


# %%
def run():
    """Hold the window open until it is closed.

    The pass ends by itself with nothing here asking: the playhead reports its
    scan ran out and the transport parks the cursor at `phrase_end` from the
    host's application clock, rather than letting it sweep off the page (rewind
    goes back to the top).
    """
    win.wait()


# %%
if __name__ == "__main__" and not hasattr(sys, "ps1"):
    try:
        run()
    finally:
        session.close()
else:
    print("score up - run() to hold the window, session.close() to end")
